### PR TITLE
Support 'untrusted' client certificate routing

### PR DESCRIPTION
**Pull request by jimini-lumox**
*Description*: Add the ability to pass through and route 'untrusted' clients.
Unit tests have not been added yet - want to get an early idea whether this PR would be rejected due to possible security concerns etc..
By default the client validation rules are unchanged.

Additional validation context options:
* request_client_certificate :   // If specified, Envoy will always request for clients to provide their certificate (without requiring it).
* permit_untrusted_client_certificate :  // If specified, Envoy will NOT reject connections using client certificates that have failed validation.
* validation_permits_no_client_certificate :  // If specified, Envoy will NOT require a client certificate if a validation context exists.
'untrusted' (either not validated, or failed validation) client certificate details added to 'x-forwarded-untrusted-client-cert' 

*Risk Level*: Medium
*Testing*: Manual so far - unit tests will be added
*Docs Changes*: API proto changes 
*Release Notes*: N/A

### CODE REVIEW

**Comment by htuch** [comment]
@jimini-lumox what's the intended use case? Unvalidated TLS is not really any more secure than plain text..

**Comment by PiotrSikora** [comment]
@htuch unverified TLS provides protection against passive observers, so it's "slightly" better than plain text :) But this PR is about unauthenticated clients in mutual TLS, and not about server certificates.

@jimini-lumox what exactly is your use case? `request_client_certificate` and `validation_permits_no_client_certificate` are already implemented by the means of [`trusted_ca`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#envoy-api-field-auth-certificatevalidationcontext-trusted-ca) and [`require_client_certificate`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#envoy-api-field-auth-downstreamtlscontext-require-client-certificate). Currently, there is no way to accept client certificates that FAILED verification, but we have [`allow_expired_certificate`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#envoy-api-field-auth-certificatevalidationcontext-allow-expired-certificate), which might work for your use case.

**Comment by htuch** [comment]
Removed stale tag, this PR should be reviewed. @PiotrSikora WDYT? I'll give this another pass in the next day.

File: api/envoy/api/v2/auth/cert.proto [hunk 1]
```diff
@@ -219,6 +219,16 @@ message CertificateValidationContext {
 
   // If specified, Envoy will not reject expired certificates.
   bool allow_expired_certificate = 8;
+
```
Comment (htuch) [inline api/envoy/api/v2/auth/cert.proto]: Can you update the description at https://github.com/envoyproxy/envoy/blob/e7d71aa3c3afe20d2a34572edd12c108f05e497d/api/envoy/api/v2/auth/cert.proto#L118 to describe how all these options interact? I share @PiotrSikora that we might have some redundancy here, in particular I think we could collapse the request/require client certificate fields with the right combination of options.

**Comment by PiotrSikora** [comment]
Oops, sorry for the delay, this felt off my radar.

I think that we could simply add `verify_client_certificate: true|check|false`, where:
- `true` would verify the client certificate and reject connection if the verification failed (what we have today),
- `check` would verify the client certificate, but forward it and the verification status via `x-forwarded-client-cert` or another header (for easier matching), regardless of the verification status,
- `false` would forward the client certificate to the backend via `x-forwarded-client-cert` without attempting to verify it (to avoid crypto operations, do access control at the backend, etc.).

I believe that this, combined with the existing [`trusted_ca`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#envoy-api-field-auth-certificatevalidationcontext-trusted-ca) and [`require_client_certificate`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#envoy-api-field-auth-downstreamtlscontext-require-client-certificate) would work for @jimini-lumox, wouldn't it?

Also, we already have the [External Authorization filter](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/ext_authz_filter), which should work even better, shouldn't it?

wrt filter chain matching, we do TLS after the filter chain and its TLS context are already selected, so doing verification, and then re-picking filter chain with a different TLS context in case of success/failure would be rather challenging and quite messy, IMHO.

**Comment by ggreenway** [comment]
+1 to what @PiotrSikora said. That's what I was thinking for this as well.

**Comment by alyssawilk** [comment]
Nice addition - probably worth including explicit release notes about it :-)

**Comment by PiotrSikora** [comment]
Moved to #5501, so that we don't forget about this feature.

**Pull request closed without merging**